neo/Implementations/Notifications/LevelDB/NotificationDB.py

```diff
@@ -166,3 +210,56 @@ def get_by_addr(self, address):
                 except Exception as e:
                     logger.info("could not parse event: %s " % val)
         return results
+
+    def get_by_contract(self, contract_hash):
+        hash = contract_hash
+        if isinstance(contract_hash, str) and len(contract_hash) == 40:
+            hash = UInt160.ParseString(contract_hash)
+
+        if not isinstance(hash, UInt160):
+            raise Exception("Incorrect address format")
+
+        contractlist_snapshot = self.db.prefixed_db(NotificationPrefix.PREFIX_CONTRACT).snapshot()
+        results = []
+
+        for val in contractlist_snapshot.iterator(prefix=bytes(hash.Data), include_key=False):
+            if len(val) > 4:
+                try:
+                    event = SmartContractEvent.FromByteArray(val)
+                    results.append(event)
+                except Exception as e:
+                    logger.info("could not parse event: %s " % val)
+        return results
+
+    def get_tokens(self):
+        """
+        Looks up all tokens
+        Returns:
+            list: A list of smart contract events with contracts that are NEP5 Tokens
+        """
+        tokens_snapshot = self.db.prefixed_db(NotificationPrefix.PREFIX_TOKEN).snapshot()
+        results = []
+        for val in tokens_snapshot.iterator(include_key=False):
+            event = SmartContractEvent.FromByteArray(val)
+            results.append(event)
+        return results
+
+    def get_token(self, hash):
+        """
+        Looks up a token by hash
+        Args:
+            hash (UInt160): The token to look up
+
+        Returns:
+            SmartContractEvent: A smart contract event with a contract that is an NEP5 Token
+        """
+        tokens_snapshot = self.db.prefixed_db(NotificationPrefix.PREFIX_TOKEN).snapshot()
+
+        try:
+            val = tokens_snapshot.get(hash.ToBytes())
+            if val:
+                event = SmartContractEvent.FromByteArray(val)
+                return event
+        except Exception as e:
+            logger.info("Smart contract event with contract hash %s not found: %s " % (hash.ToString(), e))
```

Comment:
is this an info or should it show up red like an error? Then use `logger.error` or `logger.exception`